fix(ivy): align NgModule registration timing with ViewEngine

crisbeto opened the pull request:
Currently in Ivy `NgModule` registration happens when the class is declared, however this is inconsistent with ViewEngine and requires extra generated code. These changes remove the generated code for `registerModuleFactory`, pass the id through to the `ngModuleDef` and do the module registration inside `NgModuleFactory.create`.

This PR resolves FW-1285.

mhevery reviewed (approved):
`@angular/core` LGTM for the compiler @alxhub should approve.

alxhub commented on packages/compiler-cli/src/ngtsc/annotations/src/ng_module.ts:
```diff
@@ -120,10 +120,16 @@ export class NgModuleDecoratorHandler implements DecoratorHandler<NgModuleAnalys
-    const id: Expression|null =
-        ngModule.has('id') ? new WrappedNodeExpr(ngModule.get('id') !) : null;
-
+    let id: string|DynamicValue|null = null;
```
I'm actually in favor of not attempting to interpret this value at all, and just passing it through as a `WrappedNodeExpr`, like before. The compiler doesn't need to understand this value.

alxhub commented on packages/compiler-cli/src/ngtsc/annotations/src/ng_module.ts:
```diff
@@ -29,7 +29,7 @@ export interface NgModuleAnalysis {
   declarations: Reference<ClassDeclaration>[];
   exports: Reference<ClassDeclaration>[];
-  id: Expression|null;
+  id: string|DynamicValue|null;
```
Per below, let's keep this as `Expression`.

alxhub commented on packages/compiler/src/jit_compiler_facade.ts:
```diff
@@ -91,6 +91,7 @@ export class CompilerFacadeImpl implements CompilerFacade {
       emitInline: true,
       containsForwardDecls: false,
       schemas: facade.schemas ? facade.schemas.map(wrapReference) : null,
+      id: facade.id || null,
```
We can wrap `facade.id` in a `WrappedNodeExpr` here.

crisbeto commented:
@alxhub It's been reworked based on the feedback.

alxhub commented on packages/compiler/src/compiler_facade_interface.ts:
```diff
@@ -106,6 +106,7 @@ export interface R3NgModuleMetadataFacade {
   exports: Function[];
   emitInline: boolean;
   schemas: {name: string}[]|null;
+  id: any;
```
It can remain a `string` here :)

alxhub reviewed (approved):
One comment, otherwise LGTM!

kara commented:
[presubmit](https://tap.corp.google.com/ui#id=OCL:246937836:BASE:246937916:1557192817977:30c936af)

kara commented:
This is failing Google presubmit because of a cyclical dependency:

```
angular2/rc/packages/core/src/linker/ng_module_factory_loader.ts -> 
angular2/rc/packages/core/public_api.ts -> angular2/rc/packages/core/src/core.ts -> 
angular2/rc/packages/core/src/metadata.ts -> angular2/rc/packages/core/src/metadata/directives.ts -> 
angular2/rc/packages/core/src/render3/jit/directive.ts -> 
angular2/rc/packages/core/src/render3/jit/environment.ts -> 
angular2/rc/packages/core/src/render3/index.ts -> 
angular2/rc/packages/core/src/render3/ng_module_ref.ts -> 
angular2/rc/packages/core/src/linker/ng_module_factory_loader.ts
```

alxhub commented:
We can remove `registerNgModuleType` from `jit/environment.ts` as it's no longer generated by the compiler (also remove it from `r3_identifiers.ts`). That will break the circular dependency.

crisbeto commented:
Thank you for the pointer @alxhub! This is ready for another presubmit.

kara commented:
[presubmit](http://test/OCL:246937836:BASE:247080468:1557261253192:9d545fd0)

kara commented:
Still getting TAP failures:

```
/angular2/rc/packages/core/src/linker/ng_module_factory_loader.ts -> 
/angular2/rc/packages/core/public_api.ts -> 
/angular2/rc/packages/core/src/core.ts -> 
/angular2/rc/packages/core/src/metadata.ts -> 
/angular2/rc/packages/core/src/metadata/directives.ts -> 
/angular2/rc/packages/core/src/render3/jit/directive.ts -> 
/angular2/rc/packages/core/src/render3/jit/environment.ts -> 
/angular2/rc/packages/core/src/render3/index.ts -> 
/angular2/rc/packages/core/src/render3/ng_module_ref.ts -> 
/angular2/rc/packages/core/src/linker/ng_module_factory_loader.ts
```

crisbeto committed:
fix(ivy): align NgModule registration timing with ViewEngine

Currently in Ivy `NgModule` registration happens when the class is declared, however this is inconsistent with ViewEngine and requires extra generated code. These changes remove the generated code for `registerModuleFactory`, pass the id through to the `ngModuleDef` and do the module registration inside `NgModuleFactory.create`.

This PR resolves FW-1285.

crisbeto committed:
fixup! fix(ivy): align NgModule registration timing with ViewEngine

crisbeto commented:
Moved some things around based on our discussion with @alxhub. Can you take another look? It's all in the second `fixup` commit.

kara commented:
[new presubmit](http://test/OCL:247929583:BASE:247929597:1557757256999:b9e10469)

kara commented:
merge assistance: change made to Ivy-specific API
